### ISSUE TITLE
Improvement in AV model performance over AO

### ISSUE THREAD

**Issue by lordmartian** [lordmartian/deep_avsr#5]
Thanks for your reply. 
I noted that you just implemented the transformer-CTC architecture of the paper.  I implemented the transformer-seq2seq based on Espnet, but the performance of the AV model is slightly worse than the AO model. I think it's due to context vector fusion step. 

The paper said concatenating channel-wise the context vectors of the audio and video modalities and fed to feed-forward layers. I just concatenated the two context vectors (e.g, the dimension both are (32, 123, 512)) into (32, 123, 1024) and used a linear layer into (32, 123, 512), then fed to feed-forward layers, but the performance is always slightly worse than the AO model. Do you have any idea on how to implement the channel-wise concatenation?  
I noted you used a convolution layer to fuse the audio and video embedding in the transformer-CTC. Should I use the same method for fusion of the two context vectors?

Any idea or suggestions will be appreciated. Thanks!

_Originally posted by @yuexianghubit in https://github.com/LordMartian/deep_avsr/issues/4#issuecomment-658106987_

**Comment by LordMartian** [comment]
The concatenation implementation is identical to what you have described. The convolution layer acts just as a linear layer to reduce the number of channels from 1024 to 512. So there's no difference in the channel-wise context vectors concatenation implementation. 

As described in the paper, audio-based speech recognition is more easier to learn than lip-reading and thus I think that your model is probably not using the visual information as much as it should. In order to overcome this problem, I have used three techniques:

1. Initializing the AV model decoder with the trained VO model decoder so that the AV model already knows how to lip-read and doesn't need much effort in learning it again.
2. Adding noise in the audio-modality (as described in the paper).
3. Dropping the audio modality with some probability and thus using only the video modality while AV training. 

I also think that the most difficult part in the AVSR is improving the AV model performance over the AO model. You can try the above techniques and hope that it helps to solve the problem. :smiley:

**Comment by yuexianghubit** [comment]
Great! Really thanks for your suggestions.

How about the AV model encoders?
Do I need initialise the AV encoders with the trained AO and VO encoders?

**Comment by LordMartian** [comment]
You're welcome and yes, you have to initialize with AO and VO encoders.

**Comment by LordMartian** [comment]
@yuexianghubit were you able to improve the AV model performance using the above suggestions?

**Comment by yuexianghubit** [comment]
I tried to train the AV model from scratch using your code, the issue like training VO model, both training and validation WER remain 1.0, happened. Then I initialised the AV model as you suggested, everything goes well. But I haven't got the final results yet. 

Now I am trying to use my own code based on Espnet to perform curriculum learning. because I simply combined the pretrain and train sets for training before. I think the curriculum learning will also help. After these done, I will try different training configuration.

Anyway, thanks for your suggestions!
Hope to talk about more about Audio-Visual Speech Recognition with you.

**Comment by LordMartian** [comment]
You're welcome. Do tell me whether you are able to reproduce the AV results as well and also if you are able to improve the performance of your Espnet based model.

This project was part of my master's thesis which is now complete. I won't be carrying out further research in AVSR. However,  I will be certainly glad to share the knowledge I gained during the thesis work.

**Comment by yuexianghubit** [comment]
Hi,

Have you ever tried Transformer-seq2seq framework, i am trying to build video-only model using seq2seq, but the performance i got is about 62%, far from your results 55.3% and the paper results 49.8%, Theoretically, Transformer-seq2seq should be better than transformer-CTC, that paper also indicates the seq2seq is better than CTC. I only change the model, and the data preparation and training schedule are the same as you. 

I cannot find out any reasons, do you have any idea about this situation based on your experience?
Thanks!

**Comment by LordMartian** [comment]
I just worked with the TM-CTC architecture in my project, so I don't have any practical experience with TM-seq2seq. The WER will decrease further when the AV model is used for VO results (as done in the Deep AVSR paper). This is also reflected in my results (WER decreased from 55.3% to 49.2%). A question - do you obtain the WER of 62% using greedy search or beam search (+LM)? 

Also, the authors of the Deep AVSR paper have released the code and weights of the Lip-Reading Transformer seq2seq architecture from a previous paper [here](https://github.com/afourast/deep_lip_reading) which achieves around 50% WER. The weights are for Tensorflow models so you will have to convert them. You can use them directly or as an initialization.

You can as well try to pretrain for more than 37 words and see if that gives any improvement in the performance. You can try to tweak the `LR_SCHEDULER_WAIT` and some other hyperparameter values.

Hope these suggestions help. :smiley:

**Comment by yuexianghubit** [comment]
Thanks for your suggestions.

I got the WER of 62% using beam search, but without LM.
In the Deep AVSR paper,  TM-CTC model got 54.7% with LM, is the AV model used for VO results or just VO model?  I noticed the paper said the A, V, and AV represent Audio-only, Video-only and Audio-Video model.

The validation set has no more than 22 words, I'm not sure if more words pertaining will give improvement.

Yes, I also found that code,  but I want to train the model myself for my following research.
Thanks again, I will try your suggestions.

One more thing, now I'm still running your AV model, the training is a little bit slow, even it's on 2080ti, and cannot finish in 7 days.

**Comment by LordMartian** [comment]
You can try to use LM in the beam search as well to see how much performance gain you can obtain. Although the results in the paper show that using LM doesn't give a significant improvement in the case of TM-seq2seq like in TM-CTC. Regarding the paper using the AV model for VO results, I rechecked the paper and I got mistaken about it. However, you can still try to use the AV model for VO operation, as you can see the significant reduction in WER after AV training. (You can remove the audio encoder from the AV model and you will have a VO model if that's the concern. Also, some recent papers also show that using audio information improves VO models)

In the training of my AV model, you can fix the encoders in the pretraining stage and then finetune the whole model only in the training stage. If you are doing that, then I think it should complete in about 7 days.

**Comment by yuexianghubit** [comment]
Ok, thanks!

**Comment by yuexianghubit** [comment]
Hi,
I have finished running your code, and got all the results, very close to yours.

**Comment by LordMartian** [comment]
Great! Thanks for the update.
